File: backend/internal/domain/resources.go
package domain

import (
	"time"
)

// ResourceType represents different types of resources in the game
type ResourceType string

const (
	ResourceGold ResourceType = "gold"
	ResourceMana ResourceType = "mana"
)

// Resources holds the current resource values for a player
type Resources struct {
	Gold int `json:"gold"`
	Mana int `json:"mana"`
}

// BuildingType represents different types of buildings
type BuildingType string

const (
	BuildingCommandCenter BuildingType = "command_center"
	// Future building types can be added here
)

// BuildingLevel represents the level of a building
type BuildingLevel int

const (
	Level1 BuildingLevel = 1
	Level2 BuildingLevel = 2
	Level3 BuildingLevel = 3
)

// ResourceGeneration defines how many resources a building generates
type ResourceGeneration struct {
	Gold int `json:"gold"`
	Mana int `json:"mana"`
}

// Building represents a building that can generate resources
type Building struct {
	Type              BuildingType       `json:"type"`
	Level             BuildingLevel      `json:"level"`
	PlayerIndex       int                `json:"playerIndex"`
	TopLeftRow        int                `json:"topLeftRow"`
	TopLeftCol        int                `json:"topLeftCol"`
	TurnsSinceUpgrade int                `json:"turnsSinceUpgrade"`
	LastUpgradeTime   time.Time          `json:"lastUpgradeTime"`
	ResourceGen       ResourceGeneration `json:"resourceGeneration"`
}

// GetResourceGeneration returns the resource generation for a building based on its type and level
func (b *Building) GetResourceGeneration() ResourceGeneration {
	switch b.Type {
	case BuildingCommandCenter:
		switch b.Level {
		case Level1:
			return ResourceGeneration{Gold: 3, Mana: 2}
		case Level2:
			return ResourceGeneration{Gold: 6, Mana: 4}
		case Level3:
			return ResourceGeneration{Gold: 10, Mana: 6}
		default:
			return ResourceGeneration{Gold: 3, Mana: 2}
		}
	default:
		return ResourceGeneration{Gold: 0, Mana: 0}
	}
}

// ShouldUpgrade checks if the building should upgrade based on turns passed
func (b *Building) ShouldUpgrade() bool {
	if b.Type != BuildingCommandCenter {
		return false
	}
	
	// Command centers upgrade every 3 turns
	if b.Level < Level3 && b.TurnsSinceUpgrade >= 3 {
		return true
	}
	
	return false
}

// Upgrade upgrades the building to the next level
func (b *Building) Upgrade() bool {
	if b.Level >= Level3 {
		return false
	}
	
	b.Level++
	b.TurnsSinceUpgrade = 0
	b.LastUpgradeTime = time.Now()
	b.ResourceGen = b.GetResourceGeneration()
	
	return true
}

// IncrementTurnCounter increments the turns since last upgrade
func (b *Building) IncrementTurnCounter() {
	b.TurnsSinceUpgrade++
}

// NewBuilding creates a new building
func NewBuilding(buildingType BuildingType, playerIndex, row, col int) *Building {
	b := &Building{
		Type:              buildingType,
		Level:             Level1,
		PlayerIndex:       playerIndex,
		TopLeftRow:        row,
		TopLeftCol:        col,
		TurnsSinceUpgrade: 0,
		LastUpgradeTime:   time.Now(),
	}
	b.ResourceGen = b.GetResourceGeneration()
	return b
}

// CommandCenterBuilding extends CommandCenter with building functionality
type CommandCenterBuilding struct {
	*CommandCenter
	*Building
}

// NewCommandCenterBuilding creates a command center with building functionality
func NewCommandCenterBuilding(playerIndex, topLeftRow, topLeftCol int) *CommandCenterBuilding {
	return &CommandCenterBuilding{
		CommandCenter: NewCommandCenter(playerIndex, topLeftRow, topLeftCol),
		Building:      NewBuilding(BuildingCommandCenter, playerIndex, topLeftRow, topLeftCol),
	}
}

File: lib/models/resources.dart
/// Resource models for the card game

/// Represents the resources a player has
class Resources {
  final int gold;
  final int mana;

  const Resources({
    required this.gold,
    required this.mana,
  });

  factory Resources.fromJson(Map<String, dynamic> json) {
    return Resources(
      gold: json['gold'] ?? 0,
      mana: json['mana'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'gold': gold,
      'mana': mana,
    };
  }

  Resources copyWith({
    int? gold,
    int? mana,
  }) {
    return Resources(
      gold: gold ?? this.gold,
      mana: mana ?? this.mana,
    );
  }

  @override
  String toString() => 'Resources(gold: $gold, mana: $mana)';
}

/// Represents resource generation from buildings
class ResourceGeneration {
  final int gold;
  final int mana;

  const ResourceGeneration({
    required this.gold,
    required this.mana,
  });

  factory ResourceGeneration.fromJson(Map<String, dynamic> json) {
    return ResourceGeneration(
      gold: json['gold'] ?? 0,
      mana: json['mana'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'gold': gold,
      'mana': mana,
    };
  }

  @override
  String toString() => 'ResourceGeneration(gold: $gold, mana: $mana)';
}

/// Building types
enum BuildingType {
  commandCenter,
}

/// Building levels
enum BuildingLevel {
  level1(1),
  level2(2),
  level3(3);

  final int value;
  const BuildingLevel(this.value);

  static BuildingLevel fromValue(int value) {
    switch (value) {
      case 1:
        return BuildingLevel.level1;
      case 2:
        return BuildingLevel.level2;
      case 3:
        return BuildingLevel.level3;
      default:
        return BuildingLevel.level1;
    }
  }
}

/// Represents a building that can generate resources
class Building {
  final BuildingType type;
  final BuildingLevel level;
  final int playerIndex;
  final int topLeftRow;
  final int topLeftCol;
  final int turnsSinceUpgrade;
  final DateTime lastUpgradeTime;
  final ResourceGeneration resourceGeneration;

  const Building({
    required this.type,
    required this.level,
    required this.playerIndex,
    required this.topLeftRow,
    required this.topLeftCol,
    required this.turnsSinceUpgrade,
    required this.lastUpgradeTime,
    required this.resourceGeneration,
  });

  factory Building.fromJson(Map<String, dynamic> json) {
    // Parse building type
    BuildingType type = BuildingType.commandCenter;
    if (json['type'] == 'command_center') {
      type = BuildingType.commandCenter;
    }

    // Parse building level
    BuildingLevel level = BuildingLevel.level1;
    if (json['level'] is int) {
      level = BuildingLevel.fromValue(json['level']);
    }

    // Parse last upgrade time
    DateTime lastUpgradeTime = DateTime.now();
    if (json['lastUpgradeTime'] != null) {
      try {
        lastUpgradeTime = DateTime.parse(json['lastUpgradeTime']);
      } catch (_) {
        // Ignore parse errors
      }
    }

    return Building(
      type: type,
      level: level,
      playerIndex: json['playerIndex'] ?? 0,
      topLeftRow: json['topLeftRow'] ?? 0,
      topLeftCol: json['topLeftCol'] ?? 0,
      turnsSinceUpgrade: json['turnsSinceUpgrade'] ?? 0,
      lastUpgradeTime: lastUpgradeTime,
      resourceGeneration: json['resourceGeneration'] != null
          ? ResourceGeneration.fromJson(json['resourceGeneration'])
          : const ResourceGeneration(gold: 0, mana: 0),
    );
  }

  /// Get the display name for the building level
  String get levelDisplay {
    switch (level) {
      case BuildingLevel.level1:
        return 'Level 1';
      case BuildingLevel.level2:
        return 'Level 2';
      case BuildingLevel.level3:
        return 'Level 3';
    }
  }

  /// Get the number of turns until next upgrade (3 turns per upgrade)
  int get turnsUntilUpgrade {
    if (level == BuildingLevel.level3) {
      return -1; // Max level
    }
    return 3 - turnsSinceUpgrade;
  }

  /// Check if building can upgrade
  bool get canUpgrade {
    return level != BuildingLevel.level3 && turnsSinceUpgrade >= 3;
  }

  @override
  String toString() => 'Building($type, $levelDisplay)';
}